<a href="https://colab.research.google.com/github/Sady345/Building-spatial-apps/blob/main/w02_sql_basics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SQL Basics

## Introduction

This notebook is a short introduction to SQL. It is based on the [DuckDB](https://duckdb.org/) database engine.

## Datasets

The following datasets are used in this notebook. You don't need to download them, they can be accessed directly from the notebook.

- [cities.csv](https://open.gishub.org/data/duckdb/cities.csv)
- [countries.csv](https://open.gishub.org/data/duckdb/countries.csv)

## References

- [W3Schools SQL Tutorial](https://www.w3schools.com/sql)
- [DuckDB SQL Introduction](https://duckdb.org/docs/sql/introduction.html)

## Installation

Uncomment the following cell to install the required packages.

In [ ]:
%pip install duckdb duckdb-engine jupysql
# A different way of writing sql code compared to python

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 1.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 95.2/95.2 kB 6.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 192.8/192.8 kB 13.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.8/69.8 kB 6.2 MB/s eta 0:00:00


## Library Import and Configuration

In [ ]:
import duckdb
import pandas as pd

# Import jupysql Jupyter extension to create SQL cells
%load_ext sql

Set configurations on jupysql to directly output data to Pandas and to simplify the output that is printed to the notebook.

In [ ]:
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

## Connecting to DuckDB

Connect jupysql to DuckDB using a SQLAlchemy-style connection string. You may either connect to an in memory DuckDB, or a file backed db.

In [ ]:
%sql duckdb:///:memory:
# %sql duckdb:///path/to/file.db ## % sql says all this code is sql

 If your SQL query is one line only, you may use the `%sql` magic command. For multi-line SQL query, you may use the `%%sql` magic command.

## Install extensions


Check available DuckDB extensions.

In [ ]:
%%sql

SELECT * FROM duckdb_extensions();

extension_name  loaded  installed install_path  \
0              arrow   False      False                
1       autocomplete   False      False                
2                aws   False      False                
3              azure   False      False                
4              delta   False      False                
5              excel   False      False                
6                fts   False       True   (BUILT-IN)   
7             httpfs   False      False                
8            iceberg   False      False                
9                icu    True       True   (BUILT-IN)   
10              inet   False      False                
11          jemalloc    True       True   (BUILT-IN)   
12              json    True       True   (BUILT-IN)   
13        motherduck   False      False                
14     mysql_scanner   False      False                
15           parquet    True       True   (BUILT-IN)   
16  postgres_scanner   False      False                
17           spatial   False      False                
18    sqlite_scanner   False      False                
19         substrait   False      False                
20             tpcds    True       True   (BUILT-IN)   
21              tpch    True       True   (BUILT-IN)   
22               vss   False      False                

                                          description            aliases  \
0   A zero-copy data integration between Apache Ar...                 []   
1          Adds support for autocomplete in the shell                 []   
2        Provides features that depend on the AWS SDK                 []   
3   Adds a filesystem abstraction for Azure blob s...                 []   
4                         Adds support for Delta Lake                 []   
5          Adds support for Excel-like format strings                 []   
6           Adds support for Full-Text Search Indexes                 []   
7   Adds support for reading and writing files ove...  [http, https, s3]   
8                     Adds support for Apache Iceberg                 []   
9   Adds support for time zones and collations usi...                 []   
10  Adds support for IP-related data types and fun...                 []   
11          Overwrites system allocator with JEMalloc                 []   
12                   Adds support for JSON operations                 []   
13     Enables motherduck integration with the system               [md]   
14    Adds support for connecting to a MySQL database            [mysql]   
15  Adds support for reading and writing parquet f...                 []   
16  Adds support for connecting to a Postgres data...         [postgres]   
17  Geospatial extension that adds support for wor...                 []   
18  Adds support for reading and writing SQLite da...  [sqlite, sqlite3]   
19         Adds support for the Substrait integration                 []   
20      Adds TPC-DS data generation and query support                 []   
21       Adds TPC-H data generation and query support                 []   
22  Adds indexing support to accelerate Vector Sim...                 []   

   extension_version       install_mode installed_from  
0                                  None                 
1                                  None                 
2                                  None                 
3                                  None                 
4                                  None                 
5                                  None                 
6                     STATICALLY_LINKED                 
7                                  None                 
8                                  None                 
9                     STATICALLY_LINKED                 
10                                 None                 
11                    STATICALLY_LINKED                 
12                    STATICALLY_LINKED                 
13                                 Non

DuckDB's [httpfs extension](https://duckdb.org/docs/extensions/httpfs) allows parquet and csv files to be queried remotely over http. This is useful for querying large datasets without having to download them locally. Let's install the extension and load the extension.

In [ ]:
%%sql

INSTALL httpfs;
LOAD httpfs;

Empty DataFrame
Columns: [Success]
Index: []

## Read CSV

Use the `httpfs` extension to read the `cities.csv` file from the web.

In [ ]:
%%sql

SELECT * FROM 'https://open.gishub.org/data/duckdb/cities.csv';
#selecting all from.

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

In [ ]:
%%sql

SELECT * FROM 'https://open.gishub.org/data/duckdb/countries.csv';

id            Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0      1        Afghanistan          AF         AFG             4   33.0000   
1      2            Albania          AL         ALB             8   41.0000   
2      3            Algeria          DZ         DZA            12   28.0000   
3      4     American Samoa          AS         ASM            16  -14.3333   
4      5            Andorra          AD         AND            20   42.5000   
..   ...                ...         ...         ...           ...       ...   
238  239  Wallis and Futuna          WF         WLF           876  -13.3000   
239  240     Western Sahara          EH         ESH           732   24.5000   
240  241              Yemen          YE         YEM           887   15.0000   
241  242             Zambia          ZM         ZMB           894  -15.0000   
242  243           Zimbabwe          ZW         ZWE           716  -20.0000   

     Longitude  
0         65.0  
1         20.0  
2          3.0  
3       -170.0  
4          1.6  
..         ...  
238     -176.2  
239      -13.0  
240       48.0  
241       30.0  
242       30.0  

[243 rows x 7 columns]

## Create Table

Create a table named `cities` from the `cities.csv` file.

In [ ]:
%%sql

CREATE TABLE cities AS SELECT * FROM 'https://open.gishub.org/data/duckdb/cities.csv';

Empty DataFrame
Columns: [Success]
Index: []

Create a table named `countries` from the `countries.csv` file.

In [ ]:
%%sql

CREATE TABLE countries AS SELECT * FROM 'https://open.gishub.org/data/duckdb/countries.csv';

Empty DataFrame
Columns: [Success]
Index: []

Display the table content in the database.

In [ ]:
%%sql

FROM cities;

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

In [ ]:
%%sql

FROM countries;

id            Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0      1        Afghanistan          AF         AFG             4   33.0000   
1      2            Albania          AL         ALB             8   41.0000   
2      3            Algeria          DZ         DZA            12   28.0000   
3      4     American Samoa          AS         ASM            16  -14.3333   
4      5            Andorra          AD         AND            20   42.5000   
..   ...                ...         ...         ...           ...       ...   
238  239  Wallis and Futuna          WF         WLF           876  -13.3000   
239  240     Western Sahara          EH         ESH           732   24.5000   
240  241              Yemen          YE         YEM           887   15.0000   
241  242             Zambia          ZM         ZMB           894  -15.0000   
242  243           Zimbabwe          ZW         ZWE           716  -20.0000   

     Longitude  
0         65.0  
1         20.0  
2          3.0  
3       -170.0  
4          1.6  
..         ...  
238     -176.2  
239      -13.0  
240       48.0  
241       30.0  
242       30.0  

[243 rows x 7 columns]

## The SQL SELECT statement

The `SELECT` statement is used to select data from a database. Use either `SELECT *` to select all columns, or `SELECT column1, column2, ...` to select specific columns.

`SELECT * FROM cities` is the same as `FROM cities`.

In [ ]:
%%sql

SELECT * FROM cities;

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

To limit the number of rows returned, use the `LIMIT` keyword. For example, `SELECT * FROM cities LIMIT 10` will return only the first 10 rows.

In [ ]:
%%sql

SELECT * FROM cities LIMIT 10;

id              name country  latitude  longitude  population
0   1             Bombo     UGA   0.58330   32.53330       75000
1   2       Fort Portal     UGA   0.67100   30.27500       42670
2   3           Potenza     ITA  40.64200   15.79900       69060
3   4        Campobasso     ITA  41.56300   14.65600       50762
4   5             Aosta     ITA  45.73700    7.31500       34062
5   6         Mariehamn     ALD  60.09700   19.94900       10682
6   7          Ramallah     PSE  31.90294   35.20621       24599
7   8      Vatican City     VAT  41.90001   12.44781         832
8   9           Poitier     FRA  46.58329    0.33328       85960
9  10  Clermont-Ferrand     FRA  45.77998    3.08001      233050

Select a subset of columns from the `cities` table and display the first 10 rows.

In [ ]:
%%sql

SELECT name, country FROM cities LIMIT 10;

name country
0             Bombo     UGA
1       Fort Portal     UGA
2           Potenza     ITA
3        Campobasso     ITA
4             Aosta     ITA
5         Mariehamn     ALD
6          Ramallah     PSE
7      Vatican City     VAT
8           Poitier     FRA
9  Clermont-Ferrand     FRA

To select distinct values, use the `DISTINCT` keyword. For example, `SELECT DISTINCT country FROM cities` will return only the distinct values of the `country` column.

In [ ]:
%%sql

SELECT DISTINCT country FROM cities LIMIT 10;

country
0     UGA
1     ITA
2     PSE
3     VAT
4     NAM
5     NLD
6     LIE
7     GEO
8     PRT
9     MHL

To count the number of rows returned, use the `COUNT(*)` function. For example, `SELECT COUNT(*) FROM cities` will return the number of rows in the `cities` table.

In [ ]:
%%sql

SELECT COUNT(*) FROM cities;

count_star()
0          1249

To count the number of distinct values, use the `COUNT(DISTINCT column)` function. For example, `SELECT COUNT(DISTINCT country) FROM cities` will return the number of distinct values in the `country` column.

In [ ]:
%%sql

SELECT COUNT(DISTINCT country) FROM cities;

count(DISTINCT country)
0                      200

To calculate the maximum value, use the `MAX(column)` function. For example, `SELECT MAX(population) FROM cities` will return the maximum value in the `population` column.

In [ ]:
%%sql

SELECT MAX(population) FROM cities;

max(population)
0         35676000

To calculate the total value, use the `SUM(column)` function. For example, `SELECT SUM(population) FROM cities` will return the total value in the `population` column.

In [ ]:
%%sql

SELECT SUM(population) FROM cities;

sum(population)
0     1.475535e+09

To calculate the average value, use the `AVG(column)` function. For example, `SELECT AVG(population) FROM cities` will return the average value in the `population` column.

In [ ]:
%%sql

SELECT AVG(population) FROM cities;

avg(population)
0     1.181373e+06

To order the results, use the `ORDER BY column` clause. For example, `SELECT * FROM cities ORDER BY country` will return the rows ordered by the `country` column alphabetically.

In [ ]:
%%sql

SELECT * FROM cities ORDER BY country LIMIT 10;

To order the results in descending order, use the `ORDER BY column DESC` clause. For example, `SELECT * FROM cities ORDER BY country ASC, population DESC` will return the rows ordered by the `country` column alphabetical order and then by the `population` column in descending order.

In [ ]:
%%sql

SELECT * FROM cities ORDER BY country ASC, population DESC LIMIT 10;

## The WHERE Clause

The `WHERE` clause is used to filter records. The `WHERE` clause is used to extract only those records that fulfill a specified condition.

In [ ]:
%%sql

SELECT * FROM cities WHERE country='USA'

You can use boolean operators such as `AND`, `OR`, `NOT` to filter records. For example, `SELECT * FROM cities WHERE country='USA' OR country='CAN'` will return the rows where the `country` column is either `USA` or `CAN`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country='USA' OR country='CAN';

To select US cities with a population greater than 1 million, use the following query: `SELECT * FROM cities WHERE country='USA' AND population > 1000000`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country='USA' AND population>1000000;

To select cities with the country name starting with the letter `U`, use the following query: `SELECT * FROM cities WHERE country LIKE 'U%'`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country LIKE 'U%';

To select cities with the country name ending with the letter `A`, use the following query: `SELECT * FROM cities WHERE country LIKE '%A'`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country LIKE '%A';

To select cities with the country name containing the letter `S` in the middle, use the following query: `SELECT * FROM cities WHERE country LIKE '_S_'`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country LIKE '_S_';

To select cities from a list of countries, use the `IN` operator. For example, `SELECT * FROM cities WHERE country IN ('USA', 'CAN')` will return the rows where the `country` column is either `USA` or `CAN`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country IN ('USA', 'CAN');

To select cities with a population between 1 and 10 million, use the following query: `SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000`.

In [ ]:
%%sql

SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000;

## SQL Joins

Reference: https://www.w3schools.com/sql/sql_join.asp

Here are the different types of the JOINs in SQL:

- `(INNER) JOIN`: Returns records that have matching values in both tables
- `LEFT (OUTER) JOIN`: Returns all records from the left table, and the matched records from the right table
- `RIGHT (OUTER) JOIN`: Returns all records from the right table, and the matched records from the left table
- `FULL (OUTER) JOIN`: Returns all records when there is a match in either left or right table

![](https://i.imgur.com/mITYzuS.png)

We have two sample tables: `cities` and `countries`.

There are 1,249 cities in the `cities` table and 243 countries in the `countries` table.

In [ ]:
%%sql

SELECT COUNT(*) FROM cities;

In [ ]:
%%sql

SELECT * FROM cities LIMIT 10;

In [ ]:
%%sql

SELECT COUNT(*) FROM countries;

In [ ]:
%%sql

SELECT * FROM countries LIMIT 10;

### SQL Inner Join

The `INNER JOIN` keyword selects records that have matching values in both tables. In the example, we join the `cities` table with the `countries` table using the `country` column in the `cities` table and the `Alpha3_code` column in the `countries` table. The result contains 1,244 rows, indicating that there are 5 cities that do not have a matching country.

In [ ]:
%%sql

SELECT * FROM cities INNER JOIN countries ON cities.country = countries."Alpha3_code";

Only select the `city` and `country` columns from the `cities` table and the `Country` column from the `countries` table.

In [ ]:
%%sql

SELECT name, cities."country", countries."Country" FROM cities INNER JOIN countries ON cities.country = countries."Alpha3_code";

### SQL Left Join

The `LEFT JOIN` keyword returns all records from the left table (`cities`), and the matched records from the right table (`countries`). The result contains 1,249 rows, the same number of rows as the `cities` table.

In [ ]:
%%sql

SELECT * FROM cities LEFT JOIN countries ON cities.country = countries."Alpha3_code";

### SQL Right Join

The `RIGHT JOIN` keyword returns all records from the right table (`countries`), and the matched records from the left table (`cities`). The result contains 1,291 rows.

In [ ]:
%%sql

SELECT * FROM cities RIGHT JOIN countries ON cities.country = countries."Alpha3_code";

### SQL Full Join

The `FULL JOIN` keyword returns all records when there is a match in either left or right table. The result contains 1,296 rows.

In [ ]:
%%sql

SELECT * FROM cities FULL JOIN countries ON cities.country = countries."Alpha3_code";

### SQL Union

The `UNION` operator is used to combine the result-set of two or more `SELECT` statements.

In [ ]:
%%sql

SELECT country FROM cities
UNION
SELECT "Alpha3_code" FROM countries;

## Aggregation

### Group By

The `GROUP BY` statement groups rows that have the same values into summary rows, like "find the number of cities in each country".

The `GROUP BY` statement is often used with aggregate functions (`COUNT`, `MAX`, `MIN`, `SUM`, `AVG`) to group the result-set by one or more columns.

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
ORDER BY COUNT(name) DESC;

In [ ]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
ORDER BY COUNT(name) DESC;

### Having

The `HAVING` clause was added to SQL because the `WHERE` keyword could not be used with aggregate functions.

For example, to select countries with more than 40 cities:

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

In [ ]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

## Conditional statements

The `CASE` statement goes through conditions and returns a value when the first condition is met (like an `IF-THEN-ELSE` statement). So, once a condition is true, it will stop reading and return the result. If no conditions are true, it returns the value in the `ELSE` clause.

For example, to divide cities into 3 groups based on their population:

In [ ]:
%%sql

SELECT name, population,
CASE
    WHEN population > 10000000 THEN 'Megacity'
    WHEN population > 1000000 THEN 'Large city'
    ELSE 'Small city'
END AS category
FROM cities;

## Saving results

You can save the results of a query to a new table using the `CREATE TABLE AS` statement.

In [ ]:
%%sql

CREATE TABLE cities2 AS SELECT * FROM cities;

Show the new table content.

In [ ]:
%%sql

FROM cities2;

Use the `DROP TABLE` statement to delete the table.

In [ ]:
%%sql

DROP TABLE IF EXISTS cities_usa;
CREATE TABLE cities_usa AS (SELECT * FROM cities WHERE country = 'USA');

In [ ]:
%%sql

FROM cities_usa;

Use the `INSERT INTO` statement to insert rows into a table.

In [ ]:
%%sql

INSERT INTO cities_usa (SELECT * FROM cities WHERE country = 'CAN');

## SQL Comments

Comments are used to explain sections of SQL statements, or to prevent execution of SQL statements.


### Single line comMents

Single line comments start with --.

Any text between -- and the end of the line will be ignored (will not be executed).

The following example uses a single-line comment as an explanation:

In [ ]:
%%sql

SELECT * FROM cities LIMIT 10 -- This is a comment;

### Multi-line comments

Multi-line comments start with `/*` and end with `*/`.

Any text between `/*` and `*/` will be ignored.

The following example uses a multi-line comment as an explanation:

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
/*
 * Adding Group by
 * Adding Order by
 */
GROUP BY country
ORDER BY COUNT(name) DESC
LIMIT 10;